### ISSUE TITLE
Allow variables in `function-calc-no-unspaced-operator`

### ISSUE THREAD

**Issue by MadLittleMods** [stylelint/stylelint#438]
When using [`function-calc-no-unspaced-operator`](https://github.com/stylelint/stylelint/blob/master/src/rules/function-calc-no-unspaced-operator/README.md),

```css
.foo {
   margin-top: calc(-$primary-page-heading-font-size - 2rem);
}
```

This results in:

```
2:19	⚠  Expected single space after "-" operator (function-calc-no-unspaced-operator) [stylelint]
2:28	⚠  Expected single space after "-" operator (function-calc-no-unspaced-operator) [stylelint]
2:28	⚠  Expected single space before "-" operator (function-calc-no-unspaced-operator) [stylelint]
2:33	⚠  Expected single space after "-" operator (function-calc-no-unspaced-operator) [stylelint]
2:33	⚠  Expected single space before "-" operator (function-calc-no-unspaced-operator) [stylelint]
2:41	⚠  Expected single space after "-" operator (function-calc-no-unspaced-operator) [stylelint]
2:41	⚠  Expected single space before "-" operator (function-calc-no-unspaced-operator) [stylelint]
2:46	⚠  Expected single space after "-" operator (function-calc-no-unspaced-operator) [stylelint]
2:46	⚠  Expected single space before "-" operator (function-calc-no-unspaced-operator) [stylelint]
```

---

There are two things going on here.

 - The first is the inverting `-` minus sign in front
 - The rest are from the name of the variable itself `$primary-page-heading-font-size`

We should allow plugins like [`postcss-simple-vars`](https://github.com/postcss/postcss-simple-vars) and [`postcss-css-variables`](https://github.com/MadLittleMods/postcss-css-variables) in calc statements and allow negative inverting with variable

**Comment by jeddy3** [comment]
@MadLittleMods Thanks for the suggestion and, as always, the very good write-up :)

@davidtheclark What are your thoughts on supporting this? 

* The rule _does_ already support the `var()` notation e.g. `a { top: calc(var(--red-color) + 2px); }`.  
* It doesn't support the negative inverting of the `var()` notation though. Is that even allowed in the spec?
* I don't believe we're supporting the _non standard_ syntax of _simple vars_ in any other rules.

Does this fall under our [CSS Processor](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/css-processors.md) approach, and does this make a good candiate for a `plugin` e.g. `simple-vars-function-calc-no-unspaced-operator`?

**Comment by davidtheclark** [comment]
So I think this might just involve a really simple change here: https://github.com/stylelint/stylelint/blob/master/src/rules/function-calc-no-unspaced-operator/index.js#L38 ---- considering the `-` a "sign" if it is followed immediately by a digit *or a `$`*. So unless I'm wrong, not a problem to implement.

But should we?

Given that both postcss-simple-vars and SCSS would use the `$` sign in that place, I don't mind allowing it. We would add a caveat in the docs that this is allowed. If Less users come in wanting `@` signs allowed as well, we could do that, too. (I do not that that doing `-var(...)` should work, according to the draft or the rest of CSS syntax precedent.) @MadLittleMods : Would adding the `$` sign to that check make this work with postcss-css-variables, too?

**Comment by MadLittleMods** [comment]
@davidtheclark negative inverting for custom properties, I guess, will have to be moved to a separate `negative()` or `negate()` function (name suggestions wanted).

```
margin-top: calc(negative(var(--primary-page-heading-font-size)) - 2rem);
```

**Comment by davidtheclark** [comment]
Yeah. Or just multiply by -1: `calc(calc(-1 * var(--primary-page-heading-font-size)) - 2rem);`

**Comment by MadLittleMods** [comment]
Nested `calc()` doesn't seem to work in browsers, although maybe [`postcss-calc`](https://github.com/postcss/postcss-calc) resolves correctly.

You can use parentheses it seems: `margin-left: calc((-1 * var(--primary-page-heading-font-size)) - 2rem);` which will work great.

**Comment by jeddy3** [comment]
> But should we?

I dunno. It's a tough call. Something like `postcss-simple-vars` is exactly the type of plugin that could be run before `stylelint` as it'll be a 1:1 replacement of _non standard_ syntax with _standard_ ones. Also, as you pointed out with the _less_ example, it's a slippery slope to supporting _non standard_ syntaxes.

**Comment by davidtheclark** [comment]
I wasn't thinking that the Less example showed a slippery slope --- just that there was one other variation I could think of. I honestly don't think there would be too many variable syntaxes to support here (at least not popular ones); so since it'd be easy to support the few that we know of, I think that wouldn't be a bad call. If it raises problems for us down the line, we can reconsider.

**Comment by jeddy3** [comment]
> I wasn't thinking that the Less example showed a slippery slope

Oops.

>  I think that wouldn't be a bad call. If it raises problems for us down the line, we can reconsider.

Agreed. Sounds like the most prudent course of action.
